File: dart/day9_part1_2020.dart
import 'dart:io';

void main() {
  List<int> numbers = File('input.txt').readAsLinesSync().map(int.parse).toList();

  int preambleLength = 25;
  int invalidNumber = 0;

  for (int i = preambleLength; i < numbers.length; i++) {
    bool found = false;
    for (int j = i - preambleLength; j < i - 1; j++) {
      for (int k = j + 1; k < i; k++) {
        if (numbers[j] + numbers[k] == numbers[i]) {
          found = true;
          break;
        }
      }
      if (found) break;
    }
    if (!found) {
      invalidNumber = numbers[i];
      break;
    }
  }

  print(invalidNumber);
}

File: swift/day9_part1_2020.swift

import Foundation

let input = try String(contentsOfFile: "input.txt")
let numbers = input.components(separatedBy: "\n").compactMap { Int($0) }

func findInvalidNumber(numbers: [Int], preambleLength: Int) -> Int {
    for i in preambleLength..<numbers.count {
        let preamble = Array(numbers[i - preambleLength..<i])
        var found = false
        for j in 0..<preamble.count {
            for k in j+1..<preamble.count {
                if preamble[j] + preamble[k] == numbers[i] {
                    found = true
                    break
                }
            }
            if found {
                break
            }
        }
        if !found {
            return numbers[i]
        }
    }
    return -1
}

let invalidNumber = findInvalidNumber(numbers: numbers, preambleLength: 25)
print(invalidNumber)
